clients/arweave_ipfs_s3.go

```diff
@@ -2,52 +2,99 @@ package clients
 
 import (
 	"fmt"
+	"io"
 	"net/http"
 	"net/url"
-	"strings"
 	"time"
 
-	"github.com/hashicorp/go-retryablehttp"
+	"github.com/cenkalti/backoff/v4"
+	"github.com/livepeer/catalyst-api/config"
 )
 
+const SCHEME_IPFS = "ipfs"
+const SCHEME_ARWEAVE = "ar"
+
 const MAX_COPY_DURATION = 20 * time.Minute
 
-var defaultRetryableHttpClient = newDefaultRetryableHttpClient()
-
-func newDefaultRetryableHttpClient() *http.Client {
-	client := retryablehttp.NewClient()
-	client.RetryMax = 2                          // Retry a maximum of this+1 times
-	client.RetryWaitMin = 200 * time.Millisecond // Wait at least this long between retries
-	client.RetryWaitMax = 1 * time.Second        // Wait at most this long between retries (exponential backoff)
-	client.HTTPClient = &http.Client{
-		// Give up on requests that take more than this long - the file is probably too big for us to process locally if it takes this long
-		// or something else has gone wrong and the request is hanging
-		Timeout: MAX_COPY_DURATION,
+var retryStrategy = backoff.NewConstantBackOff(1 * time.Second)
+
+func CopyDStorageToS3(url, s3URL string) error {
+	content, err := DownloadDStorageFromGatewayList(url)
+	if err != nil {
+		return fmt.Errorf("error fetching content from configured gateways: %s", err)
 	}
 
-	return client.StandardClient()
+	err = UploadToOSURL(s3URL, "", content, MAX_COPY_DURATION)
+	if err != nil {
+		return fmt.Errorf("failed to copy content to S3: %s", err)
+	}
+
+	return nil
 }
 
-func CopyArweaveToS3(arweaveURL, s3URL string) error {
-	resp, err := defaultRetryableHttpClient.Get(arweaveURL)
+func DownloadDStorageFromGatewayList(u string) (io.ReadCloser, error) {
+	var err error
+	var gateways []*url.URL
+	parsedURL, err := url.Parse(u)
 	if err != nil {
-		return fmt.Errorf("error fetching Arweave or IPFS URL: %s", err)
+		return nil, fmt.Errorf("invalid URL: %w", err)
 	}
-	defer resp.Body.Close()
 
-	err = UploadToOSURL(s3URL, "", resp.Body, MAX_COPY_DURATION)
-	if err != nil {
-		return fmt.Errorf("failed to copy Arweave or IPFS URL to S3: %s", err)
+	if parsedURL.Scheme == SCHEME_ARWEAVE {
+		gateways = config.ImportArweaveGatewayURLs
+	} else {
+		gateways = config.ImportIPFSGatewayURLs
 	}
 
-	return nil
+	var opContent io.ReadCloser
+	downloadOperation := func() error {
+		for _, gateway := range gateways {
+			path, err := url.JoinPath(gateway.Path, parsedURL.Host)
+
+			if err != nil {
+				return fmt.Errorf("cannot build gateway path: %w", err)
+			}
+
+			if gateway.Query().Has("pinataGatewayToken") {
+				values := gateway.Query()
+				values.Set("pinataGatewayToken", config.LP_PINATA_GATEWAY_TOKEN)
+				gateway.RawQuery = values.Encode()
+			}
+
+			url := url.URL{
```

Comment:
You can just use the `path` variable here (likely renaming itself to `url`, `fullURL` or sth like that)

Edit: Ohhhh, just realized now that you are using a different `JoinPath` function above. Instead of using `url.JoinPath(string, string) (string, error)`, you should use the `(*url.URL) JoinPath(string) *url.URL` instead and it will be much cleaner! It already starts from a URL and receives only some additional segments to add to the path, returning a new URL with that. It also never errs! So cleaner logic here.

The whole thing, since the beginning of the `for` loop iteration until this line can become a 1-liner:
```suggestion
			url := gateway.JoinPath(parsedURL.Host)
```

`JoinPath` will also keep the same query string of the URL, so no need to bother about that either.